File: scripts/generate_mock_data.py
#!/usr/bin/env python3
"""
Generate mock_data/cases/*.json, mock_data/index.json, and SVG thumbnails.
Deterministic (seed param). Produces 40+ cases including 10 Spanish, edge cases,
explainability, nearest_neighbors, data_quality, and provenance.
Run: python scripts/generate_mock_data.py   or: yarn gen:mock (if wired to Python)
"""
from __future__ import annotations

import argparse
import base64
import json
import os
from datetime import datetime, timezone
from pathlib import Path

import numpy as np

OUT = Path(__file__).resolve().parent.parent / "mock_data"
CASES = OUT / "cases"
THUMBS = OUT / "thumbnails"
SCHEMA_VERSION = "1.0"
MOCK_DATA_VERSION = "1.0"


def make_embedding(dim: int = 256, rng: np.random.Generator | None = None) -> str:
    rng = rng or np.random.default_rng()
    v = rng.standard_normal(size=(dim,)).astype(np.float32)
    v = v / np.linalg.norm(v)
    return base64.b64encode(v.tobytes()).decode("ascii")


def write_svg_thumb(path: Path, label: str, case_num: int) -> str:
    path.parent.mkdir(parents=True, exist_ok=True)
    svg = f'''<svg width="320" height="240" xmlns="http://www.w3.org/2000/svg">
  <rect rx="12" ry="12" width="100%" height="100%" fill="#F8F9FA"/>
  <circle cx="80" cy="80" r="40" fill="#E0F7FA"/>
  <text x="160" y="200" font-size="20" text-anchor="middle" fill="#5F6368">{label}</text>
</svg>'''
    path.write_text(svg)
    return str(path.relative_to(OUT))


# Caregiver texts and Spanish equivalents
TEXTS_EN = [
    "He says 'mama' sometimes but doesn't use two-word phrases. Points to objects.",
    "Runs and climbs well, still wobbly on one foot.",
    "Uses spoon, sometimes fork; spills often.",
    "Plays alongside others, minimal sharing.",
    "Limited eye contact, repetitive play.",
    "Parent concerned about speech.",
    "Points to things, seems social.",
    "He says 8 words and points, but no two-word phrases.",
]

TEXTS_ES = [
    "No dice palabras de dos palabras, solo balbucea.",
    "Corre y trepa bien, aún inestable en un pie.",
    "Usa cuchara, a veces tenedor; derrama a veces.",
    "Juega junto a otros, comparte poco.",
    "Contacto visual limitado, juego repetitivo.",
    "Padre preocupado por el habla.",
    "Señala cosas, parece sociable.",
    "Dice unas 8 palabras, no frases de dos palabras.",
]

RISKS = ["on_track", "monitor", "refer"]
AGES = [6, 12, 18, 24, 36, 48]
LOCALES = ["en-US", "es-ES"]


def main() -> None:
    parser = argparse.ArgumentParser(description="Generate mock cases and thumbnails")
    parser.add_argument("--seed", type=int, default=1234, help="RNG seed")
    parser.add_argument("--count", type=int, default=40, help="Number of main cases")
    parser.add_argument("--out", type=str, default=None, help="Output dir (default: mock_data)")
    args = parser.parse_args()
    if args.out:
        global OUT, CASES, THUMBS
        OUT = Path(args.out)
        CASES = OUT / "cases"
        THUMBS = OUT / "thumbnails"

    CASES.mkdir(parents=True, exist_ok=True)
    THUMBS.mkdir(parents=True, exist_ok=True)

    rng = np.random.default_rng(args.seed)
    cases_index = []

    for i in range(1, args.count + 1):
        cid = f"case-{i:04d}"
        age = int(rng.choice(AGES))
        is_spanish = i <= 10
        locale = "es-ES" if is_spanish else "en-US"
        texts = TEXTS_ES if is_spanish else TEXTS_EN
        caregiver_text = texts[i % len(texts)]

        # Risk distribution: more on_track, fewer refer
        risk = rng.choice(RISKS, p=[0.5, 0.35, 0.15])
        conf = float(np.round(rng.uniform(0.5, 0.95), 2))
        low_confidence = conf < 0.55
        if low_confidence:
            uncertainty_reason = "Limited caregiver description; consider follow-up questions."
        else:
            uncertainty_reason = None

        emb = make_embedding(256, rng)
        thumb_name = f"thumb_{i:04d}.svg"
        thumb_rel = write_svg_thumb(THUMBS / thumb_name, f"case {i}", i)

        explainability = [
            {"type": "text", "detail": "10 words reported; no two-word phrases", "score": 0.6 + rng.uniform(0, 0.3)},
        ]
        if i % 3 == 0:
            explainability.append({
                "type": "image_region",
                "detail": f"grip pattern consistent with {age}-month drawing",
                "score": 0.5 + rng.uniform(0, 0.2),
            })
        if i % 5 == 0:
            explainability.append({
                "type": "nearest_neighbor",
                "detail": "Similar to cases in training set",
                "score": 0.7,
            })

        # Nearest neighbors for a subset (FAISS-style UI)
        nearest_neighbors = []
        if i % 4 == 0:
            for j in range(1, 4):
                other = ((i + j * 7) % args.count) + 1
                if other != i:
                    nearest_neighbors.append({
                        "case_id": f"case-{other:04d}",
                        "similarity": round(0.7 + rng.uniform(0, 0.2), 2),
                        "thumbnail": f"thumbnails/thumb_{other:04d}.svg",
                    })

        mock_inference = {
            "summary": [f"Mock summary for {cid}" + (" (es)" if is_spanish else "")],
            "risk": risk,
            "recommendations": [
                "Daily language modeling activities (5–10 mins)",
                "Re-screen in 3 months",
                "Consider referral if not improving",
            ][: 2 + (risk == "refer")],
            "parent_text": (
                "Su hijo podría beneficiarse de actividades de lenguaje diarias. Revisar en 3 meses."
                if is_spanish
                else "Your child may benefit from short daily language activities. Re-screen in 3 months."
            ),
            "explainability": explainability,
            "confidence": conf,
            "adapter_id": "mock/pediscreen_v1",
            "model_id": "google/medgemma-2b-it",
        }
        if uncertainty_reason:
            mock_inference["uncertainty_reason"] = uncertainty_reason
        if nearest_neighbors:
            mock_inference["nearest_neighbors"] = nearest_neighbors

        data_quality = rng.choice(
            [["ehr_verified"], ["clinician_reviewed"], ["small_n_pilot"], ["synthetic"]],
            p=[0.2, 0.3, 0.2, 0.3],
        ).tolist()
        provenance = {
            "source": "demo" if "synthetic" in data_quality else "pilot",
            "origin": "siteA",
            "collected_by": "CHW_001",
        }

        case = {
            "schema_version": SCHEMA_VERSION,
            "case_id": cid,
            "created_at": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
            "consent_id": "consent-demo",
            "age_months": age,
            "locale": locale,
            "caregiver_text": caregiver_text,
            "images": [thumb_rel],
            "embedding_b64": emb,
            "emb_version": "medsiglip-v1",
            "mock_inference": mock_inference,
            "audit_log": [],
            "data_quality": data_quality,
            "provenance": provenance,
        }
        (CASES / f"{cid}.json").write_text(json.dumps(case, indent=2))

        cases_index.append({
            "case_id": cid,
            "age_months": age,
            "locale": locale,
            "risk": risk,
            "thumb": thumb_rel,
        })

    (OUT / "index.json").write_text(json.dumps(cases_index, indent=2))
    (OUT / "MOCK_DATA_VERSION.txt").write_text(MOCK_DATA_VERSION + "\n")
    print(f"Wrote {args.count} cases to {CASES}, index and thumbnails to {OUT}")


if __name__ == "__main__":
    main()

File: scripts/generate_mock_data.js
/**
 * Generate mock_data/cases/*.json and mock_data/index.json for PediScreen demo.
 * Run: node scripts/generate_mock_data.js  (or yarn gen:mock)
 */
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const out = path.join(__dirname, '..', 'mock_data');
const casesDir = path.join(out, 'cases');
const thumbnailsDir = path.join(out, 'thumbnails');

if (!fs.existsSync(out)) fs.mkdirSync(out, { recursive: true });
if (!fs.existsSync(casesDir)) fs.mkdirSync(casesDir, { recursive: true });
if (!fs.existsSync(thumbnailsDir)) fs.mkdirSync(thumbnailsDir, { recursive: true });

function makeEmbedding(dim = 256) {
  const arr = new Float32Array(dim);
  for (let i = 0; i < dim; i++) arr[i] = Math.random() * 2 - 1;
  const norm = Math.sqrt([...arr].reduce((s, v) => s + v * v, 0));
  for (let i = 0; i < dim; i++) arr[i] /= norm;
  return Buffer.from(arr.buffer).toString('base64');
}

const sampleTextsEn = [
  'He says 7 words, but no phrase.',
  'Points to things, seems social.',
  'Limited eye contact, repetitive play.',
  'Parent concerned about speech.',
  'He says 8 words and points, but no two-word phrases.',
  'Runs and climbs well, still wobbly on one foot.',
  'Uses spoon, sometimes fork; spills often.',
  'Plays alongside others, minimal sharing.',
];

const sampleTextsEs = [
  'No dice frases de dos palabras.',
  'Señala cosas, parece sociable.',
  'Contacto visual limitado, juego repetitivo.',
  'Padre preocupado por el habla.',
  'Dice unas 8 palabras, no frases de dos palabras.',
  'Corre y trepa bien, aún inestable en un pie.',
  'Usa cuchara, a veces tenedor; derrama a veces.',
  'Juega junto a otros, comparte poco.',
];

const risks = ['on_track', 'monitor', 'refer'];
const ages = [6, 12, 18, 24, 36, 48];

const indexList = [];

for (let i = 1; i <= 50; i++) {
  const id = `case-${String(i).padStart(4, '0')}`;
  const age = ages[Math.floor(Math.random() * ages.length)];
  const isSpanish = i <= 10;
  const locale = isSpanish ? 'es-ES' : 'en-US';
  const caregiver_text = isSpanish
    ? sampleTextsEs[Math.floor(Math.random() * sampleTextsEs.length)]
    : sampleTextsEn[Math.floor(Math.random() * sampleTextsEn.length)];

  // Edge cases: 8 cases (e.g. 41-48): missing age, noisy, preterm, etc.
  const edgeCase = i >= 41 && i <= 48;
  const adversarial = i >= 44 && i <= 50; // overlapping: ambiguous/conflicting
  const ageMonths = edgeCase && i % 4 === 0 ? null : age; // some missing age for edge

  const riskWeights = edgeCase ? [0.33, 0.34, 0.33] : adversarial ? [0.2, 0.5, 0.3] : [0.5, 0.3, 0.2];
  let ri = 0;
  let r = Math.random();
  for (let w = 0; w < riskWeights.length; w++) {
    r -= riskWeights[w];
    if (r <= 0) {
      ri = w;
      break;
    }
  }
  const risk = risks[ri];

  const mock_inference = {
    summary: [`Mock summary for ${id}${isSpanish ? ' (es)' : ''}.`],
    risk,
    recommendations: ['Short play-based activities', 'Rescreen in 3 months'],
    parent_text: isSpanish
      ? 'Pruebe actividades de lenguaje diarias; revisar en 3 meses.'
      : 'Try daily language play; re-check in 3 months.',
    explainability: [
      { type: 'text', detail: 'No two-word phrases', score: 0.6 + Math.random() * 0.3 },
    ],
    confidence: parseFloat((0.5 + Math.random() * 0.45).toFixed(2)),
    adapter_id: 'mock/adapter_v1',
    model_id: 'mock/medgemma',
  };

  const doc = {
    schema_version: '1.0',
    case_id: id,
    created_at: new Date().toISOString(),
    consent_id: 'consent-demo',
    age_months: ageMonths,
    locale,
    caregiver_text,
    images: [`thumbnails/thumb_${String(i).padStart(3, '0')}.svg`],
    embedding_b64: makeEmbedding(),
    emb_version: 'medsiglip-v1',
    mock_inference,
    audit_log: i % 5 === 0 ? [
      {
        timestamp: new Date(Date.now() - 3600000).toISOString(),
        actor_type: 'model',
        actor_id: 'mock/medgemma',
        action: 'inference',
        reason: '',
        model_version: 'mock-v1',
        adapter_id: 'mock/adapter_v1',
      },
    ] : [],
  };

  if (i <= 15) {
    doc.time_to_referral_days = risk === 'refer' ? 7 + Math.floor(Math.random() * 14) : null;
    doc.D_base = risk !== 'on_track' && Math.random() < 0.4;
  }
  doc.data_quality = ['synthetic'];
  doc.provenance = { source: 'demo', origin: 'siteA', collected_by: 'CHW_001' };
  if (mock_inference.confidence < 0.55) mock_inference.uncertainty_reason = 'Limited caregiver description; consider follow-up questions.';
  if (i % 4 === 0) {
    mock_inference.explainability.push({ type: 'image_region', detail: `grip pattern consistent with ${(ageMonths ?? age)}-month drawing`, score: 0.5 + Math.random() * 0.2 });
  }
  if (i % 5 === 0) {
    mock_inference.nearest_neighbors = [
      { case_id: `case-${String((i % 50) + 1).padStart(4, '0')}`, similarity: 0.75 + Math.random() * 0.15, thumbnail: `thumbnails/thumb_${String((i % 50) + 1).padStart(3, '0')}.svg` },
    ];
  }

  fs.writeFileSync(path.join(casesDir, `${id}.json`), JSON.stringify(doc, null, 2));

  indexList.push({
    case_id: doc.case_id,
    age_months: doc.age_months,
    locale: doc.locale,
    risk: doc.mock_inference.risk,
    thumb: doc.images[0],
  });
}

fs.writeFileSync(
  path.join(out, 'index.json'),
  JSON.stringify(indexList, null, 2)
);
fs.writeFileSync(path.join(out, 'MOCK_DATA_VERSION.txt'), '1.0\n');

// Placeholder SVG thumbnails (no real photos)
for (let i = 1; i <= 50; i++) {
  const name = `thumb_${String(i).padStart(3, '0')}.svg`;
  const svg = `<svg xmlns="http://www.w3.org/2000/svg" width="120" height="120" viewBox="0 0 120 120"><rect width="120" height="120" fill="#f0f0f0"/><text x="60" y="65" text-anchor="middle" fill="#999" font-size="14">${i}</text></svg>`;
  fs.writeFileSync(path.join(thumbnailsDir, name), svg);
}

console.log('Wrote 50 mock cases to mock_data/cases/ and mock_data/index.json');
console.log('Wrote placeholder thumbnails to mock_data/thumbnails/');
